gspread/worksheet.py

```diff
@@ -542,40 +542,165 @@ def get_all_records(
+        values_len = len(values[0])
+        keys_len = len(keys)
+        values_wider_than_keys_by = values_len - keys_len
+        default_blank_in_keys = default_blank in keys
+
+        if ((values_wider_than_keys_by > 0) and default_blank_in_keys) or (
+            values_wider_than_keys_by > 1
+        ):
+            raise GSpreadException(
+                "the header row in the worksheet contains multiple empty cells"
+            )
+        elif values_wider_than_keys_by == 1:
+            keys.append(default_blank)
+        elif values_wider_than_keys_by < 0:
+            values = [row + [""] * (-values_wider_than_keys_by) for row in values]
+
+        return values, keys
+
+    def get_records_subset(
+        self,
+        empty2zero=False,
+        head=1,
+        first_row=None,
+        last_row=None,
+        default_blank="",
+        allow_underscores_in_numeric_literals=False,
+        numericise_ignore=[],
+        value_render_option=None,
+        expected_headers=None,
+    ):
+        """Returns a list of dictionaries, all of them having the contents of
+        the spreadsheet range selected with the head row as keys and each of these
+        dictionaries holding the contents of subsequent selected rows of cells as
+        values.
+
+        Cell values are numericised (strings that can be read as ints or floats
+        are converted), unless specified in numericise_ignore
+
+        :param bool empty2zero: (optional) Determines whether empty cells are
+            converted to zeros.
+        :param int head: (optional) Determines which row to use as keys,
+            starting from 1 following the numeration of the spreadsheet.
+        :param int first_row: (optional) row to start reading data from (inclusive) (1-based).
+        :param int last_row: (optional) row to stop reading at (inclusive) (1-based).
+        :param str default_blank: (optional) Determines which value to use for
+            blank cells, defaults to empty string.
+        :param bool allow_underscores_in_numeric_literals: (optional) Allow
+            underscores in numeric literals, as introduced in PEP 515
+        :param list numericise_ignore: (optional) List of ints of indices of
+            the columns (starting at 1) to ignore numericising, special use
+            of ['all'] to ignore numericising on all columns.
+        :param value_render_option: (optional) Determines how values should
+            be rendered in the output. See `ValueRenderOption`_ in
+            the Sheets API.
+        :type value_render_option: :namedtuple:`~gspread.utils.ValueRenderOption`
+
+        :param list expected_headers: (optional) List of expected headers, they must be unique.
+
+            .. note::
+
+                returned dictionaries will contain all headers even if not included in this list
+
+        """
+        # some sanity checks
+        head, first_row, last_row = self._validate_rows_ranges_for_get_records_subset(
```

Comment:
these should be done in-place. no need to define extra functions for them

Reply:
resolving this comment as https://github.com/burnash/gspread/pull/1301#discussion_r1334639018 is similar, will keep the discussion there.